File: Data Structures & Algorithms/trapping-rain-water/submission-3.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l, r =[0] * n, [0] * n
        l_wall = 0
        r_wall = 0
        count = 0

        for i in range(n):
            j = -i - 1
            l[i] = l_wall
            r[j] = r_wall
            l_wall = max(l_wall, height[i])
            r_wall = max(r_wall, height[j])

        
        for i in range(n):
            pots = min(l[i], r[i])
            count += max(0, pots - height[i])
        
        return count
            

File: Data Structures & Algorithms/perfect-squares/submission-3.go
func numSquares(n int) int {
	dp := make([]int, n + 1)

    for i := range n + 1 {
        dp[i] = n
    }

    dp[0] = 0

	for i:= 1; i < n + 1; i++ {
		for j := 1; j < i + 1; j++ {
            sq := j * j

            if i - sq < 0 {
                break
            }

			dp[i] = min(dp[i], 1 + dp[i - sq])
		}
	}
    
    return dp[n]
}
